### PR TITLE
refactor recall_precision metric for better DDP support

### LINKED ISSUE #1073: box_recall and box_precision don't match inside training_loop and outside
I have not been able to recreate this issue locally yet, but I want to document in case a user sees it. I believe this will only occur on the pre2.0 tag on the git repo and not on any release. 

In https://www.comet.com/bw4sz/boem/f3c687c8824f4b9fafa4bafbbe3fad8b?compareXAxis=step&experiment-tab=panels&prevPath=%2Fbw4sz%2Fboem%2Fview%2Fnew%2Fexperiments&showOutliers=true&smoothing=0&xAxis=step
the box_recall and precision is 0.8 and 0.15 on the logged metrics, but afterwards when running main.evaluate on the same code you get the expect high score of 0.91 and 0.92. A couple posibilities

This likely relates to the on_validation_epoch_end hook which has been a consistant challenge because of the inability to use trainer.predict inside a hook that is called during trainer.fit, which is well documented within pytorch lightning. 

# Potential ideas.

1. The metrics aren't being logged currently
2. The predictions aren't being formatted correctly.
3. The metrics and predictions are being passed correctly, but the model isn't being called correctly, batch_norm and any batch statistics.

### LINKED ISSUE #1232: Evaluation reports spuriously low recall if limit_batches is set
Was doing some test runs with a limited batch size and the results look much better than the box recall suggests (< 0.03, which is low, even for a bad model). Evaluate loads the ground truth CSV file and doesn't use the dataloader, so it will run over the whole val dataset and assume missing predictions for images that were skipped. (I suspect this anyway).

This would probably fix it:

```python
        if self.trainer.limit_val_batches != 1.0 and self.val_dataloader() != []:
            val_loader = self.val_dataloader()
            valid_image_paths = set()
            for batch in val_loader:
                _, _, image_names = batch
                valid_image_paths.update(image_names)
            ground_df = ground_df[ground_df.image_path.isin(valid_image_paths)]
```

but seems wasteful to do a full loop over the dataloader.

This is useful for easily testing the effect of dataset size on model performance, but we would probably want to generate specific subset train and val files to be more rigorous.

### PR DESCRIPTION
**Pull request by jveitchmichaelis**
## Description

- Refactors the RecallPrecision metric to use (mostly) the same `update` signature as other metrics.
- Ground truth/pred matching is performed eagerly during `update` so that the final reduction stage doesn't time-out the sync between ranks during DDP joining.
- This should be faster on multi node machines, because each rank will only perform 1/N of the work. There are maybe some tiny benefits from not having to search the dataframe for image_path <> preds/gt. The `it/s` during eval might be a bit slower, but there won't be a big chunk of compute at the end.
- The metric has to keep track of the matching results so it can be accessed later, which is a little awkward, but I've tried to keep it clean (at least to users...)
- I moved empty frame accuracy here too, since we have to handle those checks during evaluation anyway and it make main a bit cleaner.
- The metrics are calculated against the dataset targets directly (instead of loading the CSV and passing that raw). This has the benefit that adding geometric augmentation to the validation dataset (if we did that for some reason) wouldn't break eval.

Tests:
- Added a sanity check in `test_main::test_evaluate` to confirm that the output dataframe is what we expect + some comments on the assertions.
- One fixup in tests for the new metric.update signature.

## Related Issue(s)

- Fixes https://github.com/weecology/DeepForest/issues/1073
- Should also fix https://github.com/weecology/DeepForest/issues/1232

## AI-Assisted Development

Rewrote most of the metric and then got CC to check some of the logic and the test suite.

- [X] I used AI tools (e.g., GitHub Copilot, ChatGPT, etc.) in developing this PR
- [X] I understand all the code I'm submitting
- [X] I have reviewed and validated all AI-generated code

**AI tools used (if applicable):**
Claude Code Opus/Sonnet 4.6

### CODE REVIEW

**Review by bw4sz — COMMENTED** [review]
This looks right, i'm going to check it out and submit a BOEM detection multi-gpu that was previous failing and see.

**Review by bw4sz — APPROVED** [review]
Agreed. I'm glad you checked out a different branch. I'm not surprised there is something specific about the bird detector.
